### PR TITLE
Optimizer refactor

### PR DESCRIPTION
**Pull request by twicki**
## Technical Description

Refactor the optimizer to handle pass groups

### Resolves / Enhances

3rd bullet in https://github.com/MeteoSwiss-APN/dawn/issues/558

### CODE REVIEW

**Comment by twicki** [comment]
launch jenkins

**Comment by twicki** [comment]
launch jenkins

**Comment by twicki** [comment]
launch jenkins

File: dawn/src/dawn/Compiler/DawnCompiler.cpp [hunk 1]
```diff
@@ -179,93 +242,175 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
...
+  //===-----------------------------------------------------------------------------------------
+  if(shouldRunGroup(options_, options_.MergeStages)) {
+    // merging requires the stage graph
+    optimizer.pushBackPass<PassSetStageGraph>();
+    // running the actual pass
+    optimizer.pushBackPass<PassStageMerger>();
+    // since this can change the scope of temporaries ...
+    optimizer.pushBackPass<PassTemporaryType>();
+    optimizer.pushBackPass<PassLocalVarType>();
+    // this pass is broken as hell... Johann please
```
Comment (jdahm) [inline dawn/src/dawn/Compiler/DawnCompiler.cpp]: Needs to change before merge

File: dawn/src/dawn/Optimizer/OptimizerOptions.inc [hunk 2]
```diff
@@ -19,75 +19,97 @@
...
 //
 
-OPT(int, MaxHaloPoints, 3, "max-halo", "", 
+OPT(int, MaxHaloPoints, 3, "max-halo", "",
     "Set the maximum number of allowed halo points", "<N>", true, false)
 OPT(int, block_size_i, 0, "block-size-i", "", "i block size for tiled computations", "", true, false)
 OPT(int, block_size_j, 0, "block-size-j", "", "j block size for tiled computations", "", true, false)
 OPT(int, block_size_k, 0, "block-size-k", "", "k block size for tiled computations", "", true, false)
+
+OPT(bool, DefaultNone, false, "default-none", "",
```
Comment (jdahm) [inline dawn/src/dawn/Optimizer/OptimizerOptions.inc]: Options will be reduced in a future PR.

**Pull request merged**